cassiopeia/solver/SharedMutationJoiningSolver.py

```diff
@@ -0,0 +1,298 @@
+"""
+This file stores the SharedMutationJoiningSolver. The inference procedure is 
+an agglomerative clustering procedure that joins samples that share the most
+identical character/state mutations.
+"""
+
+from typing import Callable, Dict, List, Optional, Tuple, Union
+
+import abc
+import networkx as nx
+import numba
+import numpy as np
+import pandas as pd
+
+from cassiopeia.data import CassiopeiaTree
+from cassiopeia.data import utilities as data_utilities
+from cassiopeia.solver import CassiopeiaSolver, solver_utilities
+
+
+class SharedMutationJoiningSolverError(Exception):
+    """An Exception class for SharedMutationJoiningSolver."""
+
+    pass
+
+
+class SharedMutationJoiningSolver(CassiopeiaSolver.CassiopeiaSolver):
```

Comment:
Is the only reason why we don't want to build on the `DistanceSolver` API because we don't want to overwrite the `dissimilarity_map` attribute in a tree?

The code in this class looks almost _identical_ to the code in the `DistanceSolver` classes, barring some name changes (e.g. dissimilarity -> similarity). If the only concern is that you don't want to overwrite the dissimilarity map, then there are three things we can do:

(1) Take your advice from last week and _not_ overwrite the dissimilarity map in a tree if you pass in a dissimilarity function.
(2) Copy the `CassiopeiaTree` where you can overwrite the dissimilarity map with a similarity map and then populate the original `CassiopeiaTree` with the resulting tree.
(3) Just run `utilities.compute_dissimilarity_map` on the character matrix in the original `CassiopeiaTree`. As opposed to `cassiopeia_tree.compute_dissimilarity_map`, this won't overwrite anything and will just get the dissimilarity map.

Reply:
I think that the main reason that I didn't want to use the DistanceSolver API is because in addition to just changing just the dissimilarity map like in UPGMA and in NJ, we also need to update and query a working character matrix as we introduce new LCA nodes from collapsing. I couldn't think of a good way to incorporate this into the DistanceSolver framework without it being clunky.

Reply:
I see your point. I don't see a right way to do this either, so we'll just have to live with duplicated code for now. Thanks for clarifying!